### DISCUSSION TITLE
How to deal with Microsoft.AspNetCore.Http.Abstractions while migrating from 2.2 to 3.1?

### DISCUSSION THREAD

**Discussion by lesscodetxm** [dotnet/aspnetcore#23276] [General]
We have a lot of ASP.Net APIs that we're migrating to 3.1
We created a (.Net Standard 2.0) core shared library that offers a bunch of conveniences and extension methods to handle service registrations, Polly policies, etc. That library references `Microsoft.AspNetCore.Http` (2.2.2) for abstractions like `IApplicationBuilder` and `RequestDelegate`.
It looks like these now live in `Microsoft.AspNetCore.Http.Abstractions`, but since there's no 3.1 version of that assembly available the advice seems to be to add a framework reference to `Microsoft.AspNetCore.App` and change the `TargetFramework` to `netcoreapp3.1`.

Unfortunately our shared library is consumed by other types of apps (eg. full framework 4.7.1 WPF apps), so we can't really do that. If we set the `TargetFramework` to `netcoreapp3.1;netstandard2.0` the compiler fails to find the Asp.Net namespace and abstractions (but I doubt that would solve our problem anyway).

I assume we need to separate out our core library into two assemblies, right? Or is there a simpler solution?

**Comment by javiercn** [comment]
@lesscodetxm thanks for contacting us.

I believe you are right and the only solution to this is break it into multiple assemblies. That said, even if you do that, you can keep the source in one place and reference it through LinkBase, something like `<Compile Include="..\..\Code\**\*" LinkBase=".\" />`

**Reply by javiercn** [reply]
@davidfowl in case he has some other idea

**Reply by davidfowl** [reply]
You don't need 2 assemblies, you can use a single assemblies and ifdefs. You can add the framework reference in the netcoreapp3.1 target. There's guidance here https://docs.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-3.1&tabs=visual-studio#migrate-libraries-via-multi-targeting
